### ISSUE TITLE
Customizing in-place accumulation

### ISSUE THREAD

**Issue by ararslan** [JuliaDiff/ChainRules.jl#39]
In porting Nabla's reverse mode sensitivity definitions to ChainRules, I've found that it's not particularly straightforward to port those which directly update a value on the tape. The equivalent operation in ChainRules appears to be `accumulate!(value, rule, args...)`, which accumulates the result of `rule(args...)` to `value`. However, most rules are defined in terms of anonymous functions which are themselves defined within `frule`/`rrule`, which means we can't specialize `accumulate!` on particular rules in the outer scope.

I see a few options here:
1. Move rule definitions which have updating counterparts to separate functions, so that one can write
   ```julia
   dfdx(x) = something
   rrule(f, x) = (f(x), Rule(dfdx))
   accumulate!(value, ::Rule{typeof(dfdx)}, x) = whatever
   ```
   One downside to this is that evaluation of the rule and accumulating the result to an existing value can't share any intermediate steps, which could mean using more memory. Another downside is that it would require reshuffling all of the current definitions outside of `frule` and `rrule` into a bunch of top-level named functions.

   On the other hand, an upside to this is that we could precompile the functions defined at the top level, which would likely help with #35.

2. Have `Rule`s store a second function alongside the sensitivity propagation function which can be used for in-place updating. Then we would have
   ```julia
   struct Rule{F<:Function,U<:Union{Function,Nothing}} <: AbstractRule
       f::F
       u::U
   end
   Rule(f) = Rule{typeof(f),Nothing}(f, nothing)
   ```
   `accumulate!` would then use the `u` function for updating if it isn't `nothing`, otherwise it would fall back to the current definition in terms of `materialize!`. The function `u` would have a signature like `u(value, args...)` which accumulates `f(args...)` to `value`.

3. Something else entirely.

I kind of like option 2, though 1 seems fine as well, if a bit suboptimal in terms of the work required to implement that change up front. Thoughts?

**Comment by willtebbutt** [comment]
I like option 2 quite a lot. Option 1 does seem rather sub-optimal in terms of implementation burden and the ease with which intermediate computations can be cached.

@jrevels any thoughts?

**Comment by jrevels** [comment]
Yup option 2 sounds like a good solution

**Comment by oxinabox** [comment]
Another option.
If i didn't misunderstand.

Would be to add a type-param to 
`Rule` that is the `typeof` the function that you are implementing the sensitivity for.
And one would manually specify that when constructing a `Rule`
Then things could be the same as Nabla, I think?

This would also solve a problem Wessel mentioned to me today, that it the function you are doing the sensititivity for is not mentioned in the stacktrack when things error.

**Comment by willtebbutt** [comment]
I like this idea even more tbh.

> This would also solve a problem Wessel mentioned to me today, that it the function you are doing the sensititivity for is not mentioned in the stacktrack when things error.

This is a _very_ good point.

**Comment by ararslan** [comment]
I think that's a good idea to do anyway, but I don't see how it addresses the concern about in-place updating here. Can you elaborate?

**Comment by oxinabox** [comment]
@ararslan I didn't fully understand Option 1, but I thought this was  similar.

**Comment by ararslan** [comment]
Oh, so you would implement the custom accumulation by specializing on
```julia
accumulate!(value, ::Rule{typeof(f)}, args...)
```
? Is that what you're thinking?

**Comment by oxinabox** [comment]
Yes

**Comment by willtebbutt** [comment]
Having thought about this a bit more, I think I'm pro- option 2 again. From a performance perspective it's something of a no-brainer. There's still the issue of stack traces though. Perhaps we introduce an additional type parameter for `Rule` as suggested by @oxinabox to make it possible to figure out what's `Rule` comes from what `rrule` / `frule` invocation?

**Comment by jrevels** [comment]
Yeah, I think Alex's option 2 is the right way to do this, since it allows the rule author to easily share state/computation between closures created for the returned `Rule`. 

> Would be to add a type-param to Rule that is the typeof the function that you are implementing the sensitivity for.

Also had this at one point in the distant past, but got rid of it since I couldn't think of a use case for it. Better stacktraces seems likes a good reason, though.

**Comment by ararslan** [comment]
Cool. I have option 2 implemented in #42. I can make the change to include the function in a separate PR, since it's a bit tangential to #42.
